src/main/java/org/opensearch/neuralsearch/processor/TextEmbeddingProcessor.java

```diff
@@ -0,0 +1,269 @@
+    Map<String, Object> buildVectorOutput(
+        Map<String, Object> knnMap,
+        List<List<Float>> modelTensorList,
+        Map<String, Object> sourceAndMetadataMap
+    ) {
+        IndexWrapper indexWrapper = new IndexWrapper(0);
+        Map<String, Object> result = new LinkedHashMap<>();
+        for (Map.Entry<String, Object> knnMapEntry : knnMap.entrySet()) {
+            String knnKey = knnMapEntry.getKey();
+            Object sourceValue = knnMapEntry.getValue();
+            if (sourceValue instanceof String && StringUtils.isNotBlank(sourceValue.toString())) {
+                List<Float> modelTensor = modelTensorList.get(indexWrapper.index++);
+                result.put(knnKey, modelTensor);
+            } else if (sourceValue instanceof List) {
+                result.put(knnKey, processListOut((List<String>) sourceValue, modelTensorList, indexWrapper));
+            } else if (sourceValue instanceof Map) {
+                processMapTypeVectorOutput(knnKey, sourceValue, modelTensorList, indexWrapper, sourceAndMetadataMap);
+            }
+        }
+        return result;
+    }
+
+    @SuppressWarnings({ "unchecked" })
+    private void processMapTypeVectorOutput(
+        String knnKey,
+        Object sourceValue,
+        List<List<Float>> modelTensorList,
+        IndexWrapper indexWrapper,
+        Map<String, Object> sourceAndMetadataMap
+    ) {
+        if (knnKey == null || sourceAndMetadataMap == null || sourceValue == null) return;
+        if (sourceValue instanceof Map) {
+            for (Map.Entry<String, Object> inputNestedMapEntry : ((Map<String, Object>) sourceValue).entrySet()) {
+                processMapTypeVectorOutput(
+                    inputNestedMapEntry.getKey(),
+                    inputNestedMapEntry.getValue(),
+                    modelTensorList,
+                    indexWrapper,
+                    (Map<String, Object>) sourceAndMetadataMap.get(knnKey)
+                );
+            }
+        } else if (sourceValue instanceof String) {
+            sourceAndMetadataMap.put(knnKey, modelTensorList.get(indexWrapper.index++));
+        } else if (sourceValue instanceof List) {
+            sourceAndMetadataMap.put(knnKey, processListOut((List<String>) sourceValue, modelTensorList, indexWrapper));
+        }
+    }
+
+    private List<Map<String, List<Float>>> processListOut(
+        List<String> sourceValue,
+        List<List<Float>> modelTensorList,
+        IndexWrapper indexWrapper
+    ) {
+        List<Map<String, List<Float>>> numbers = new ArrayList<>();
+        for (String strSourceValue : sourceValue) {
+            if (StringUtils.isNotBlank(strSourceValue)) {
+                numbers.add(ImmutableMap.of(LIST_TYPE_NESTED_MAP_KEY, modelTensorList.get(indexWrapper.index++)));
+            }
+        }
+        return numbers;
+    }
+
+    private static void validateEmbeddingFieldsType(IngestDocument ingestDocument, Map<String, Object> embeddingFields) {
```

Comment:
any reason for making this and other functions static? if no please make them no static.

Reply:
Static functions means this method do not depends on any stateful data, so other methods can use this without concerns of the state changing. Please let me know your thoughts on making it no static. Thanks.

Reply:
So, this function is not dependent on any stateful correct that is correct, but it is always used in the context of a Stateful and that too only in this class.

So, seems like with your assumption if a function is not dependent on a state it should always be static. That seems to be an overkill. Static function lives in JVM heap and generally created to keep data in heap and make them run faster. We don't have any of the case here.

If this function was present in another class then we might want to make this function static as we don't want to create objects at runtime as object creation is expensive. But that is not the case here.

Reply:
> Static function lives in JVM heap and generally created to keep data in heap and make them run faster. We don't have any of the case here.
If this function was present in another class then we might want to make this function static as we don't want to create objects at runtime as object creation is expensive. But that is not the case here.

The reducing of time complexity is the benefit we can get from a static method instead of the unnecessary of object creation, please refer to: https://stackoverflow.com/a/135038.  This running faster itself is a benefit for us.